File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    const dp = {}
    for (let i = 0; i <= cost.length; ++i) {
        dp[i] = -1;
    }
    
    return Math.min(traverse(cost, dp, cost.length-1), traverse(cost, dp, cost.length-2));
};

/**
 * @param {number} cost
 * @param {number} dp
 * @return {number}
 */
var traverse = (cost, dp, index) => {
    if (index < 0) return 0;
    if (dp[index] >= 0) return dp[index];
    dp[index] = cost[index] + Math.min(traverse(cost, dp, index-1), traverse(cost, dp, index-2));
    
    return dp[index];
}

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution 
{
private:
    
public:
    int minCostClimbingStairs(vector<int>& cost) 
    {
        if (cost.size() == 2) 
            return min(cost[0], cost[1]); 
        for (int i = 2; i < cost.size(); ++i)
        {
            cost[i] = min(cost[i] + cost[i-1], cost[i] + cost[i-2]);
        }
        return min(cost[cost.size()-1], cost[cost.size()-2]);
    }

};

/*
cost = [1, 100, 1, 1, 1, 100, 1, 1, 100, 1]
1 100 1 1 1 100 1 1 100 1
1 100 2 3 3 103 4 5 104 6
*/